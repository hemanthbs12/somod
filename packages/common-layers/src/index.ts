import { existsSync } from "fs";
import { dirname, join } from "path";

// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace layerLibraries {
  export const base: string[] = [
    "@solib/json-validator",
    "@solib/common-types-schemas",
    "@solib/errors",
    "lodash",
    "tslib",
    "uuid"
  ];
  export const customResource: string[] = ["@solib/cfn-custom-resource"];
  export const httpWrapper: string[] = ["@solib/http-lambda"];
}

export const getLocation = async (): Promise<string> => {
  let location = __dirname;
  while (!existsSync(join(location, "package.json"))) {
    location = dirname(location);
  }
  return location;
};
