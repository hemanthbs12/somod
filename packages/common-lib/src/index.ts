export * from "./json-validator";
export * from "./schema";
export * from "./errors";
