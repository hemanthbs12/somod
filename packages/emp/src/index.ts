import { rootCommand } from "@sodaru-cli/base";
import buildCommand from "./commands/build";
import deployCommand from "./commands/deploy";
import initCommand from "./commands/init";
import serveCommand from "./commands/serve";

const program = rootCommand("emp", [
  initCommand,
  buildCommand,
  serveCommand,
  deployCommand
]);

export default program;
