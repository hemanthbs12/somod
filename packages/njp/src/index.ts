import { rootCommand } from "@solib/cli-base";
import buildCommand from "./commands/build";
import serveCommand from "./commands/serve";
import initCommand from "./commands/init";
import prepareCommand from "./commands/prepare";

const program = rootCommand("njp", [
  initCommand,
  buildCommand,
  prepareCommand,
  serveCommand
]);

export default program;
