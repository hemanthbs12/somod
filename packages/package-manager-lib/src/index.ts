export * from "./tasks/packageJson";

export * from "./tasks/ignoreFiles";

export * from "./tasks/tsConfigBuildJson";

export * from "./tasks/init";

export * from "./utils/constants";
