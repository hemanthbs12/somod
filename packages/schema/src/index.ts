export * from "./common/keywords";
export * from "./serverless-template/components/expression";
export * from "./serverless-template/components/keywords";
export * from "./serverless-template/resources/all";
export * from "./serverless-template/resources/custom";
export * from "./serverless-template/resources/function";
export * from "./serverless-template/resources/functionLayer";
export * from "./ui-config/index";
export { default as parameters } from "./parameters/index";
export { default as serverless } from "./serverless-template/index";
export { default as uiConfig } from "./ui-config/index";
