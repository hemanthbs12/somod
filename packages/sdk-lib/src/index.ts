export * from "./findRootDir";

export * from "./tasks/packageJson";

export * from "./tasks/ignoreFile";

export * from "./tasks/tsConfigBuildJson";

export * from "./tasks/common";

export * from "./tasks/nextJs";

export * from "./tasks/serverless";

export * from "./tasks/parameters";

export * from "./utils/constants";
