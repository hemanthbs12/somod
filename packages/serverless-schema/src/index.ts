import Ajv, { AnySchemaObject, ValidateFunction } from "ajv";
import { readFile } from "fs/promises";
import { get as httpGet } from "http";
import { get as httpsGet } from "https";
import { join } from "path";
import { URL } from "url";
import { isPlainObject, isArray, isString } from "lodash";

const idBase = "https://json-schema.sodaru.com";

const loadFromFile = async (path: string): Promise<string> => {
  const content = readFile(path, { encoding: "utf8" });
  return content;
};

const loadFromWeb = (uri: string): Promise<string> => {
  const requestFunction = uri.startsWith("https://") ? httpsGet : httpGet;
  return new Promise((resolve, reject) => {
    const request = requestFunction(uri, res => {
      if (!(res.statusCode >= 200 && res.statusCode < 300)) {
        reject(
          new Error(
            `Invalid response from ${uri}, statusCode : ${res.statusCode}`
          )
        );
      }
      const chunks: string[] = [];
      res.on("data", chunk => {
        chunks.push(chunk);
      });
      res.on("end", () => {
        resolve(chunks.join(""));
      });
    });
    request.on("error", e => {
      reject(e);
    });
  });
};

const getModuleFromUri = (uri: string): string => {
  let moduleName = null;
  if (uri.startsWith(idBase)) {
    const url = new URL(uri);
    const paths = url.pathname.split("/");
    if (paths.length >= 2) {
      moduleName = paths[1].startsWith("@")
        ? paths[1] + "/" + paths[2]
        : paths[1];
    }
  }
  if (!moduleName) {
    throw new Error(`Could not find modulename from ${uri}`);
  }
  return moduleName;
};

const uriReferenceToUri = (input: string, base: string): string => {
  let _input = input;
  if (_input.indexOf("/node_modules/")) {
    const baseModuleName = getModuleFromUri(base) as string;
    const replacement = baseModuleName.startsWith(`@`) ? "/../../" : "/../";
    _input = _input.replace("/node_modules/", replacement);
  }
  const url = new URL(_input, base);
  return url.toString();
};

const extractReferences = (
  schema: string
): { schema: AnySchemaObject; $refs: string[] } => {
  const correctedSchema = schema.replaceAll(
    "https://json-schema.org/draft-07/schema", // vs-code extension redhat.vscode-yaml requires https://json-schema.org/draft-07/schema
    "http://json-schema.org/draft-07/schema" // ajv requires http://json-schema.org/draft-07/schema
  );

  const schemaObj: AnySchemaObject = JSON.parse(correctedSchema);
  schemaObj.$schema = uriReferenceToUri(schemaObj.$schema, schemaObj.$id);

  const $_refs: string[] = [];
  const objectsToBeParsed = [schemaObj];
  while (objectsToBeParsed.length > 0) {
    const objectToBeParsed = objectsToBeParsed.shift();
    if (isPlainObject(objectToBeParsed)) {
      if (isString(objectToBeParsed.$ref)) {
        const $_ref = uriReferenceToUri(objectToBeParsed.$ref, schemaObj.$id);
        objectToBeParsed.$ref = $_ref;
        $_refs.push($_ref);
      }
      objectsToBeParsed.push(...Object.values(objectToBeParsed));
    } else if (isArray(objectToBeParsed)) {
      objectsToBeParsed.push(...objectToBeParsed);
    }
  }
  const $refs = $_refs
    .map($ref => {
      const url = new URL($ref);
      return `${url.protocol}//${url.host}${url.pathname}`; // drop fragment in uri
    })
    .filter($ref => $ref != schemaObj.$id); // filter references to same file

  return { schema: schemaObj, $refs };
};

const visitedIds: string[] = [];

const resetVisitedIds = () => {
  // RESET visited Ids
  while (visitedIds.length > 0) {
    visitedIds.pop();
  }
  visitedIds.push("http://json-schema.org/draft-07/schema");
};

const parseAndLoad = async (
  id: string,
  ajv: Ajv,
  dir: string,
  rootModuleName: string,
  loadType: "addSchema" | "addMetaSchema" | "compile"
): Promise<void | ValidateFunction> => {
  if (!visitedIds.includes(id)) {
    visitedIds.push(id);

    let fileContent: string = null;
    if (id.startsWith(idBase)) {
      const moduleName = getModuleFromUri(id);
      const moduleRootDir =
        moduleName == rootModuleName
          ? dir
          : join(dir, "node_modules", moduleName);
      const relativePath = id.substr(`${idBase}/${moduleName}/`.length);
      fileContent = await loadFromFile(join(moduleRootDir, relativePath));
    } else {
      fileContent = await loadFromWeb(id);
    }

    const { schema, $refs } = extractReferences(fileContent);
    await parseAndLoad(
      schema.$schema,
      ajv,
      dir,
      rootModuleName,
      "addMetaSchema"
    );
    await Promise.all(
      $refs.map($ref =>
        parseAndLoad($ref, ajv, dir, rootModuleName, "addSchema")
      )
    );

    if (loadType == "addMetaSchema") {
      if (!ajv.getSchema(id)) {
        ajv.addMetaSchema(schema);
      }
    } else if (loadType == "addSchema") {
      if (!ajv.getSchema(id)) {
        ajv.addSchema(schema);
      }
    } else {
      return ajv.compile(schema);
    }
  }
};

export const compile = async (
  dir: string,
  ajv: Ajv
): Promise<ValidateFunction> => {
  const packageJsonStr = await readFile(join(dir, "package.json"), {
    encoding: "utf8"
  });
  const packageJson = JSON.parse(packageJsonStr);
  const id = packageJson.serverlessSchema;

  const thisModuleName = "@somod/serverless-schema";

  const schemaToCompile =
    id || `${idBase}/${thisModuleName}/schemas/index.json`;

  resetVisitedIds();

  const validateFunction = (await parseAndLoad(
    schemaToCompile,
    ajv,
    dir,
    packageJson.name,
    "compile"
  )) as ValidateFunction;

  return validateFunction;
};
