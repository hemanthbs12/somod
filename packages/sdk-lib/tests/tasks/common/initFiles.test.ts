/* eslint-disable no-console */
import { createFiles, createTempDir, deleteDir } from "@sodev/test-utils";
import { listFiles } from "@solib/cli-base";
import { initFiles } from "../../../src";
import { yellow } from "chalk";
import { readFile } from "fs/promises";
import { join } from "path";

describe("Test task initFiles", () => {
  let dir: string = null;
  const originalWarn = console.warn;

  beforeEach(() => {
    dir = createTempDir();
    console.warn = jest.fn();
  });

  afterEach(() => {
    console.warn = originalWarn;
    deleteDir(dir);
  });

  test("only common", async () => {
    const result = await initFiles(dir, false, false);
    expect(result).toBeUndefined();
    const downloadedFiles = await listFiles(dir);
    expect(downloadedFiles).toMatchSnapshot();
    expect(console.warn).toHaveBeenCalledTimes(0);
  });

  test("only ui", async () => {
    const result = await initFiles(dir, true, false);
    expect(result).toBeUndefined();
    const downloadedFiles = await listFiles(dir);
    expect(downloadedFiles).toMatchSnapshot();
    expect(console.warn).toHaveBeenCalledTimes(0);
  });

  test("only serverless", async () => {
    const result = await initFiles(dir, false, true);
    expect(result).toBeUndefined();
    const downloadedFiles = await listFiles(dir);
    expect(downloadedFiles).toMatchSnapshot();
    expect(console.warn).toHaveBeenCalledTimes(0);
  });

  test("for empty dir", async () => {
    const result = await initFiles(dir, true, true);
    expect(result).toBeUndefined();
    const downloadedFiles = await listFiles(dir);
    expect(downloadedFiles).toMatchSnapshot();
    expect(console.warn).toHaveBeenCalledTimes(0);
  });

  test("for non-empty dir", async () => {
    createFiles(dir, { "lib/index.ts": "" });
    const result = await initFiles(dir, true, true);
    expect(result).toBeUndefined();
    expect(console.warn).toHaveBeenCalledTimes(1);
    expect(console.warn).toHaveBeenCalledWith(
      yellow("lib/index.ts exists : SKIPPED")
    );
    await expect(
      readFile(join(dir, "lib/index.ts"), { encoding: "utf8" })
    ).resolves.toEqual("");
  });

  test("for forcing replacement in non-empty dir", async () => {
    createFiles(dir, { "lib/index.ts": "" });
    const result = await initFiles(dir, true, true, true);
    expect(result).toBeUndefined();
    expect(console.warn).toHaveBeenCalledTimes(1);
    expect(console.warn).toHaveBeenCalledWith(
      yellow("lib/index.ts exists : REPLACED")
    );
    await expect(
      readFile(join(dir, "lib/index.ts"), { encoding: "utf8" })
    ).resolves.not.toEqual("");
  });
});
