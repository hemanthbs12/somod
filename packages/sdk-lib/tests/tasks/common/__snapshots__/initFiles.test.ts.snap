// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test task initFiles for empty dir 1`] = `
Array [
  "lib/index.ts",
  "parameters.yaml",
  "serverless/template.yaml",
  "tsconfig.json",
  "ui/config.yaml",
  "ui/pages/_document.tsx",
  "ui/pages/index.tsx",
  "ui/public/favicon.ico",
]
`;

exports[`Test task initFiles only common 1`] = `
Array [
  "lib/index.ts",
]
`;

exports[`Test task initFiles only serverless 1`] = `
Array [
  "lib/index.ts",
  "serverless/template.yaml",
]
`;

exports[`Test task initFiles only ui 1`] = `
Array [
  "lib/index.ts",
  "parameters.yaml",
  "tsconfig.json",
  "ui/config.yaml",
  "ui/pages/_document.tsx",
  "ui/pages/index.tsx",
  "ui/public/favicon.ico",
]
`;
